File: client/nodejs/rsp_net.js
'use strict';

// RSP socket abstraction layer for Node.js.
//
// Provides RSPSocket and RSPServer as streaming replacements for the BSD sockets
// the C++ client creates. RSPSocket is a Duplex stream compatible with the Node.js
// net module API. RSPServer is an EventEmitter that behaves like net.Server.
//
// Usage:
//   const {createConnection, createServer} = require('./rsp_net');
//
//   // Server side
//   const server = await createServer(client, rmNodeId, '127.0.0.1:8080');
//   server.on('connection', (socket) => {
//       socket.on('data', (chunk) => socket.write(chunk)); // echo
//   });
//
//   // Client side
//   const socket = await createConnection(client, rmNodeId, '127.0.0.1:8080');
//   socket.write('hello');
//   socket.on('data', (chunk) => console.log(chunk.toString()));

const {Duplex, EventEmitter} = require('stream');
const messages = require('./messages');

const {SUCCESS, SOCKET_DATA, SOCKET_CLOSED, SOCKET_ERROR, NEW_CONNECTION} = messages.SOCKET_STATUS;
const TRACE = process.env.RSP_NET_TRACE === '1';

function trace(message) {
    if (TRACE) {
        console.error(`[rsp_net] ${message}`);
    }
}

// RSPSocket wraps a single RSP socket ID as a Node.js Duplex stream.
//
// Data received from the remote end is pushed as Buffer chunks via the readable
// side. Data written to the writable side is sent via the RSP socket. The socket
// is closed when the stream is ended or destroyed.
class RSPSocket extends Duplex {
    constructor(client, socketId) {
        super({allowHalfOpen: false});
        this._client = client;
        this._socketId = socketId;
        this._closing = false;

        client.attachSocketHandler(socketId, (reply) => this._onSocketReply(reply));
    }

    get socketId() {
        return this._socketId;
    }

    _onSocketReply(reply) {
        const status = reply.error || 0;
        trace(`socket=${this._socketId} reply_status=${status}`);
        if (status === SOCKET_DATA) {
            const data = reply.data ? Buffer.from(reply.data, 'hex') : Buffer.alloc(0);
            trace(`socket=${this._socketId} data_len=${data.length}`);
            this.push(data);
        } else if (status === SOCKET_CLOSED || status === SOCKET_ERROR) {
            this._closing = true;
            this._client.detachSocketHandler(this._socketId);
            this.push(null);
        }
        // SUCCESS replies from fire-and-forget sends are silently ignored.
    }

    _write(chunk, encoding, callback) {
        if (this._closing) {
            callback(new Error('socket is closed'));
            return;
        }
        const data = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk, encoding);
        trace(`socket=${this._socketId} write_len=${data.length}`);
        this._client.socketSend(this._socketId, data)
            .then((sent) => {
                if (!sent) {
                    callback(new Error(`socket send failed for socket ${this._socketId}`));
                    return;
                }
                callback();
            })
            .catch(callback);
    }

    _read(_size) {
        // Data is pushed asynchronously from _onSocketReply; nothing to pull here.
    }

    _final(callback) {
        // Preserve half-close semantics: allow reading remote response data
        // after local writes end (e.g., HTTP request followed by response).
        callback();
    }

    _destroy(err, callback) {
        this._doClose().catch(() => {}).finally(() => callback(err));
    }

    _doClose() {
        if (this._closing) return Promise.resolve();
        this._closing = true;
        this._client.detachSocketHandler(this._socketId);
        return this._client.socketClose(this._socketId)
            .then((closed) => {
                if (!closed) {
                    trace(`socket=${this._socketId} close not acknowledged`);
                }
            });
    }
}

// RSPServer wraps a listening RSP socket and emits 'connection' events with
// RSPSocket instances for each incoming connection.
//
// The server uses async-accept mode: the resource manager delivers NEW_CONNECTION
// replies automatically without requiring explicit acceptTCP calls.
class RSPServer extends EventEmitter {
    constructor(client, nodeId, listenSocketId) {
        super();
        this._client = client;
        this._nodeId = nodeId;
        this._listenSocketId = listenSocketId;
        this._closed = false;

        client.attachSocketHandler(listenSocketId, (reply) => this._onListenReply(reply));
    }

    get socketId() {
        return this._listenSocketId;
    }

    _onListenReply(reply) {
        const status = reply.error || 0;
        if (status === NEW_CONNECTION && reply.new_socket_id?.value) {
            // _handleSocketReply already registered the route for new_socket_id.
            const socket = new RSPSocket(this._client, reply.new_socket_id.value);
            this.emit('connection', socket);
        } else if (status === SOCKET_CLOSED || status === SOCKET_ERROR) {
            if (!this._closed) {
                this._closed = true;
                this._client.detachSocketHandler(this._listenSocketId);
                this.emit('close');
            }
        }
    }

    async close() {
        if (this._closed) return;
        this._closed = true;
        this._client.detachSocketHandler(this._listenSocketId);
        await this._client.socketClose(this._listenSocketId).catch(() => {});
    }
}

// Create an RSPSocket connected to hostPort via the given RSP node.
// Returns a connected RSPSocket ready for reading and writing.
async function createConnection(client, nodeId, hostPort, options = {}) {
    const reply = await client.connectTCPEx(nodeId, hostPort, {asyncData: true, ...options});
    const status = reply?.error ?? SUCCESS;
    if (status !== SUCCESS || !reply?.socket_id?.value) {
        throw Object.assign(new Error(`RSP connect failed (status=${status})`), {status, reply});
    }
    return new RSPSocket(client, reply.socket_id.value);
}

// Create an RSPServer listening on hostPort via the given RSP node.
// Returns an RSPServer that emits 'connection' for each incoming RSPSocket.
async function createServer(client, nodeId, hostPort, options = {}) {
    const reply = await client.listenTCPEx(nodeId, hostPort, {
        asyncAccept: true,
        childrenAsyncData: true,
        ...options,
    });
    const status = reply?.error ?? SUCCESS;
    if (status !== SUCCESS || !reply?.socket_id?.value) {
        throw Object.assign(new Error(`RSP listen failed (status=${status})`), {status, reply});
    }
    return new RSPServer(client, nodeId, reply.socket_id.value);
}

module.exports = {RSPSocket, RSPServer, createConnection, createServer};

File: client/python/rsp_net.py
"""Async RSP socket abstraction — Python equivalent of client/nodejs/rsp_net.js."""

import asyncio
import base64

from rsp_client import RSPClient, SUCCESS, STREAM_DATA, STREAM_CLOSED, STREAM_ERROR, NEW_CONNECTION


def _stream_id_value(reply: dict | None) -> str | None:
    if not isinstance(reply, dict):
        return None
    return (
        (reply.get("stream_id") or {}).get("value")
        or (reply.get("socket_id") or {}).get("value")
    )


def _new_stream_id_value(reply: dict | None) -> str | None:
    if not isinstance(reply, dict):
        return None
    return (
        (reply.get("new_stream_id") or {}).get("value")
        or (reply.get("new_socket_id") or {}).get("value")
    )


class RSPSocket:
    """Wraps a single RSP socket ID. Use as an async context manager."""

    def __init__(self, client: RSPClient, socket_id: str) -> None:
        self._client = client
        self._socket_id = socket_id
        self._recv_queue: asyncio.Queue[bytes | None] = asyncio.Queue()
        self._closed = False
        client.attach_socket_handler(socket_id, self._on_socket_reply)

    @property
    def socket_id(self) -> str:
        return self._socket_id

    def _on_socket_reply(self, reply: dict) -> None:
        status = reply.get("error") or 0
        if status == STREAM_DATA:
            data_b64 = reply.get("data") or ""
            data = base64.b64decode(data_b64) if data_b64 else b""
            self._recv_queue.put_nowait(data)
        elif status in (STREAM_CLOSED, STREAM_ERROR):
            self._closed = True
            self._client.detach_socket_handler(self._socket_id)
            self._recv_queue.put_nowait(None)

    async def read(self, n: int = -1) -> bytes:
        """Read the next data chunk. Returns b'' on EOF."""
        if self._closed and self._recv_queue.empty():
            return b""
        chunk = await self._recv_queue.get()
        return chunk if chunk is not None else b""

    async def write(self, data: bytes) -> None:
        """Send data using fire-and-forget send."""
        sent = await self._client.socket_send(self._socket_id, data)
        if not sent:
            raise ConnectionError(f"socket send failed for socket {self._socket_id}")

    async def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        self._client.detach_socket_handler(self._socket_id)
        await self._client.socket_close(self._socket_id)

    async def __aenter__(self):
        return self

    async def __aexit__(self, *args):
        await self.close()

    def __aiter__(self):
        return self

    async def __anext__(self) -> bytes:
        data = await self.read()
        if not data:
            raise StopAsyncIteration
        return data


class RSPServer:
    """Async RSP server. Emits connections via accept() or async iteration."""

    def __init__(self, client: RSPClient, node_id: str, listen_socket_id: str) -> None:
        self._client = client
        self._node_id = node_id
        self._listen_socket_id = listen_socket_id
        self._closed = False
        self._accept_queue: asyncio.Queue[RSPSocket | None] = asyncio.Queue()
        client.attach_socket_handler(listen_socket_id, self._on_listen_reply)

    @property
    def socket_id(self) -> str:
        return self._listen_socket_id

    def _on_listen_reply(self, reply: dict) -> None:
        status = reply.get("error") or 0
        if status == NEW_CONNECTION:
            new_stream_id = _new_stream_id_value(reply)
            if new_stream_id:
                socket = RSPSocket(self._client, new_stream_id)
                self._accept_queue.put_nowait(socket)
        elif status in (STREAM_CLOSED, STREAM_ERROR):
            if not self._closed:
                self._closed = True
                self._client.detach_socket_handler(self._listen_socket_id)
                self._accept_queue.put_nowait(None)

    async def accept(self) -> "RSPSocket":
        """Wait for and return the next incoming RSPSocket."""
        if self._closed and self._accept_queue.empty():
            raise ConnectionError("server is closed")
        sock = await self._accept_queue.get()
        if sock is None:
            raise ConnectionError("server closed")
        return sock

    async def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        self._client.detach_socket_handler(self._listen_socket_id)
        await self._client.socket_close(self._listen_socket_id)

    async def __aenter__(self):
        return self

    async def __aexit__(self, *args):
        await self.close()

    def __aiter__(self):
        return self

    async def __anext__(self) -> "RSPSocket":
        try:
            return await self.accept()
        except ConnectionError:
            raise StopAsyncIteration


async def create_connection(
    client: RSPClient, node_id: str, host_port: str, **options
) -> RSPSocket:
    """Connect to host_port via the RSP node and return an RSPSocket."""
    reply = await client.connect_tcp_ex(node_id, host_port, async_data=True, **options)
    status = (reply.get("error") if reply else None) or SUCCESS
    stream_id = _stream_id_value(reply)
    if status != SUCCESS or not stream_id:
        raise ConnectionError(f"RSP connect failed (status={status})")
    return RSPSocket(client, stream_id)


async def create_server(
    client: RSPClient, node_id: str, host_port: str, **options
) -> RSPServer:
    """Listen on host_port via the RSP node and return an RSPServer."""
    reply = await client.listen_tcp_ex(
        node_id, host_port,
        async_accept=True,
        children_async_data=True,
        **options,
    )
    status = (reply.get("error") if reply else None) or SUCCESS
    stream_id = _stream_id_value(reply)
    if status != SUCCESS or not stream_id:
        raise ConnectionError(f"RSP listen failed (status={status})")
    return RSPServer(client, node_id, stream_id)
